### ISSUE TITLE
SparseTensor incompatible with ParameterBuffer/GradientTape autodiff pipeline

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#98]
## Bug Description

`SparseTensor<T>` inherits from `Tensor<T>` to participate in the autodiff tape system, but `ParameterBuffer<T>.CopyFrom()` and `FlattenGradients()` call `Tensor<T>.Contiguous()` which throws `InvalidOperationException` for sparse tensors via `ThrowIfSparse()`. This blocks any neural network with sparse parameters (e.g., `SparseNeuralNetwork`, `SparseLinearLayer`) from using `TrainWithTape()`.

## Reproduction

```csharp
var network = new SparseNeuralNetwork<double>(...);
var input = new Tensor<double>([1, 16]);
var target = new Tensor<double>([1, 8]);
network.Train(input, target);  // Throws InvalidOperationException
```

## Error

```
System.InvalidOperationException : Contiguous is not supported on sparse tensors.
Use SparseTensor-specific APIs or call ToDense() first.

Stack Trace:
  at TensorBase.ThrowIfSparse()                          // TensorBase.cs:524
  at Tensor<T>.Contiguous()                              // Tensor.cs:150
  at ParameterBuffer<T>.CopyFrom(IReadOnlyList<Tensor<T>>)  // ParameterBuffer.cs:176
  at NeuralNetworkBase<T>.GetOrCreateParameterBuffer()   // NeuralNetworkBase.cs:2626
  at NeuralNetworkBase<T>.TrainWithTape()                // NeuralNetworkBase.cs:2494
```

### Call chain (two failure points)

**1. `ParameterBuffer<T>.CopyFrom()` (line 176):**
```csharp
var contiguous = param.IsContiguous ? param : param.Contiguous();
```
When `param` is a `SparseTensor<T>`, `IsContiguous` returns `false` (sparse storage isn't contiguous), so it calls `Contiguous()` which hits `ThrowIfSparse()` at `Tensor.cs:150`.

**2. `ParameterBuffer<T>.FlattenGradients()` (line 229):**
```csharp
var contiguous = grad.IsContiguous ? grad : grad.Contiguous();
```
Same issue — if gradient tensors are sparse, this path also throws.

### Why SparseTensor.IsContiguous is false

`SparseTensor<T>` inherits `IsContiguous` from `Tensor<T>`, which checks if strides match row-major order. Since sparse tensors use index arrays (COO/CSR/CSC), not strided dense storage, `IsContiguous` returns false. But even if it returned true, accessing `.DataVector` on a sparse tensor returns only the **non-zero values**, not the full dense data — so the span copy at line 179 would produce wrong results anyway.

### The design intent

`SparseTensor<T> : Tensor<T>` was designed so sparse tensors participate in the autodiff tape as first-class citizens (matching PyTorch's design where `torch.sparse` tensors have `requires_grad=True` and flow through autograd). However, the `ParameterBuffer` assumes all parameter tensors have dense contiguous storage, breaking this.

## Proposed Fix

The fix should be in `ParameterBuffer<T>`, not in `SparseTensor` — the buffer should handle sparse parameters by densifying them during copy. This matches PyTorch's approach where sparse gradients are coalesced/densified when accumulated into the parameter buffer.

### Option A: Densify sparse tensors in CopyFrom/FlattenGradients (recommended)

```csharp
// ParameterBuffer.CopyFrom, line ~176
Tensor<T> dense;
if (param.IsSparse)
{
    dense = ((SparseTensor<T>)param).ToDense();
}
else
{
    dense = param.IsContiguous ? param : param.Contiguous();
}
var srcData = dense.DataVector;
// ... rest unchanged
```

Same pattern for `FlattenGradients` at line ~229.

**Pros:** Minimal change, SparseTensor.ToDense() already exists and is tested.
**Cons:** Allocates dense copy every training step (same as PyTorch's sparse grad accumulation).

### Option B: Implement Contiguous() on SparseTensor

Override `Contiguous()` in `SparseTensor<T>` to return `ToDense()`:

```csharp
// SparseTensor<T>
public override Tensor<T> Contiguous()
{
    return ToDense();  // Returns dense [Rows, Columns] tensor
}
```

And remove the `ThrowIfSparse()` guard in `Tensor<T>.Contiguous()`, replacing it with a virtual dispatch.

**Pros:** More transparent — any code calling `.Contiguous()` works automatically.
**Cons:** Requires changing the base `Tensor<T>.Contiguous()` from non-virtual to virtual, which is a broader API change.

### Option C: Sparse-aware ParameterBuffer

Add a sparse parameter path that keeps parameters sparse in the buffer and only densifies for the optimizer step.

**Pros:** Best performance for high-sparsity models.
**Cons:** Significant complexity increase.

## Recommendation

**Option A** is the safest fix for v0.27.x — it's a 4-line change in `ParameterBuffer.cs` with no API changes. Option B is better for v0.28+ as a principled API improvement.

## Affected Components

- `ParameterBuffer<T>.CopyFrom()` — line 176
- `ParameterBuffer<T>.FlattenGradients()` — line 229
- All networks using `TrainWithTape()` with sparse layers (`SparseNeuralNetwork`, `SparseLinearLayer`)

## Version

AiDotNet.Tensors v0.27.1
